### ISSUE TITLE
Feature: Try guessing Entrypoints 

### ISSUE THREAD

**Issue by frank-dspeed** [rollup/rollup#3664]
### Feature Use Case
supply a glob pattern and then guess entrypoints
https://github.com/TomerAberbach/find-entry-points

### Feature Proposal
Good question at present i have no clear view i expect that creating a virtual entrypoint called app or something like that would work that imports every JS file so that is the main app chunk and then we could maybe manual mark or create chunks in the generate bundle hook?

**Comment by frank-dspeed** [comment]
@lukastaegert as your known for good ideas i would be happy if you could look into that.

**Comment by TomerAberbach** [comment]
@frank-dspeed, thanks for creating this issue!

@lukastaegert, here's some plugin code I wrote to illustrate what I'm trying to achieve here:
```js
import { basename } from 'path'
import { findEntryPoints } from 'find-entry-points'
import globby from 'globby'

const autoEntry = () => ({
  name: `auto-entry`,
  async options (options) {
    const { input, ...rest } = options

    return {
      ...rest,
      input: await globby(input)
    }
  },
  async generateBundle (options, bundle) {
    const entryPoints = new Set(
      (await findEntryPoints(this.getModuleIds(), {
        parseImports: ({ file: { path: moduleId } }) => {
          const { importedIds, dynamicallyImportedIds } = this.getModuleInfo(moduleId)
          return importedIds.concat(dynamicallyImportedIds)
        }
      })).map(entryPoints =>
        entryPoints.find(entryPoint => basename(entryPoint) === `index.js`) ?? entryPoints[0]
      )
    )

    for (const fileName of Object.keys(bundle)) {
      if (bundle[fileName].type === `chunk`) {
        delete bundle[fileName]
      }
    }

    for (const entryPoint of entryPoints) {
      this.emitFile({
        type: `chunk`,
        id: entryPoint
      })
    }
  }
})

export default autoEntry
```
I'm essentially allowing Rollup to do the work of constructing the dependency graph by expanding the glob and setting all of the files as entry points. However, I ultimately want to designate only some files as entry points based on the return value of `findEntryPoints`. I attempted to do this by deleting all of the chunks, and then emitting only the files designated as entry points by `findEntryPoints`, but it does not work because I cannot use `this.emitFile` in the `generateBundle` hook.

I tried all sorts of ideas (big thanks to @frank-dspeed for helping there), but I cannot replicate the behavior of setting the entry points directly. I think it would be really helpful if I could somehow designate chunks as entry points or not entry points after they have been created.

I would rather not construct the dependency graph on my own prior to running Rollup (and invoking `findEntryPoints` to get the entry points to pass to Rollup from the beginning) because I'm afraid there will be subtle differences between the dependency graph that I construct and the one that Rollup would have constructed, especially if the user is using plugins that define `load` or `resolve` hooks; in this case my dependency graph may be very different.

Looking forward to hearing your thoughts!

**Comment by frank-dspeed** [comment]
## little note only internal 

- maybe the new caching API can help with that as it would allow us to run analyze more often so we could generate the graph or edit it better and more often.
- maybe we could introduce something like incremental Graph creation and validation via Refactoring a lot of OO Stuff to function Composition.

**Comment by TomerAberbach** [comment]
Hey @lukastaegert! Just wanted to follow up on this issue. Let me know if you need me to clarify anything!

**Comment by lukastaegert** [comment]
Honestly, it does not feel like a feature that is useful to many to me since
- as a library, you do not have explicit control over your public interface
- as a web developer, you do not have explicit control over your load order
- circular dependency handling will result in many entry points where a human could easily make a decision which is the "correct" one
So I am rather sceptical about putting something like this into core.

That being said, what you could do is just run a second Rollup instance from the plugin `buildStart` hook. You would just need the first phase and then in the `buildEnd` hook, you can do your graph analysis. Then emit the entry points you want and resolve the buildStart hook. If a `cache` is provided, you can get it in the `options` hook of the original run and pass it to the second run for speed-up.

**Comment by TomerAberbach** [comment]
Hey @lukastaegert, thanks for the feedback!

I agree that this functionality is not well-suited for core because, as you have noted, it is very opinionated. Not sure if that was @frank-dspeed's intention when creating this issue (my goal is to make a plugin for this).

I tried something similar to what you suggested, and it does seem to work. I'm going to play around with it a bit more (to see how it handles edge cases). If it works well, then I'll close this issue.

**Comment by lukastaegert** [comment]
I thought a little more about it, and I think there is actually something we could do to make writing such a plugin much easier. So the thing is, assuming the module graph would already exist in some way, then we already have a mechanism to designate entry points, which is `this.emitFile({type: 'chunk', ...})`. However, you would first need a mechanism to add modules to the graph without designating entry points.
- So the first bit would be adding a new context method, e.g. `this.addModule(id)`. This would put `id` directly through the `load` and `transform` hooks to load and parse it, add it to the graph and then also resolve, load and parse all static and dynamic dependencies and add them to the graph. So it would be similar to `this.emitFile` except that it does not make the module an entry point. This could also be useful for the CommonJS plugin to trigger scanning a module to find out if it is ESM or CommonJS.
- Then you need to emit the entries when the graph is complete. The graph is complete in the `buildEnd` hook, but you cannot emit chunks in that hook because, well, then the build would not have ended. So the second bit would be a new hook, let's call it `modulesLoaded` or `graphComplete` or something better we can come up with. It would be called each time the module loader runs idle because all dependencies have been added to the graph. Normally, this would just happen once directly before `buildEnd`. But the advantage would be that it would allow you to emit additional chunks in that hook, which may or may not trigger loading additional modules. If new modules are added, the hook is called again. Only if no additional modules or chunks are added in the hook by any plugin, we would proceed to `buildEnd`.

So these are just ideas. I hope you understand that from my side, they are low priority at the moment but if you want to start on a PR for one of the other, that would be very welcome and I would like to support you.

**Comment by TomerAberbach** [comment]
Thanks for giving this more thought!

Perhaps I don't fully understand your proposal, but I think even with your suggested functionality the plugin remains difficult to implement as long as Rollup requires that `input` is specified in the input options. At the moment, I have the plugin written like so:
```js
import { basename } from 'path'
import { rollup } from 'rollup'
import { findEntryPoints } from 'find-entry-points'
import globby from 'globby'

const name = `auto-entry`

const rollupForEntryPoints = async ({ input, plugins, ...rest }) => {
  let entryPoints

  const autoEntryInternal = {
    name: `${name}-internal`,
    async buildEnd() {
      entryPoints = (
        await findEntryPoints(this.getModuleIds(), {
          parseImports: ({ file: { path: moduleId } }) => {
            const { importedIds, dynamicallyImportedIds } = this.getModuleInfo(
              moduleId
            )
            return importedIds.concat(dynamicallyImportedIds)
          }
        })
      ).map(
        entryPointGroups =>
          entryPointGroups.find(
            entryPoint => basename(entryPoint) === `index.js`
          ) ?? entryPointGroups[0]
      )
    }
  }

  const bundle = await rollup({
    ...rest,
    input: await globby(input),
    plugins: [autoEntryInternal].concat(
      (plugins ?? []).filter(plugin => plugin.name !== name)
    )
  })

  return {
    cache: bundle.cache,
    entryPoints
  }
}

const autoEntry = () => ({
  name,
  async options(options) {
    const { cache, entryPoints } = await rollupForEntryPoints(options)

    return {
      ...options,
      cache,
      input: entryPoints
    }
  }
})
```

Earlier you suggested that I run a second Rollup instance in the `buildStart` hook, but I can't do that because I would need to specify the `input` option for Rollup, and I wouldn't know what to specify. If I specify the expanded glob (i.e. all of the JS files), then every file would be considered an entry point, and I have no way to mark them as non-entry points later.

So I don't think your suggested changes would make the plugin easier to implement unless we introduced a way to mark entry points as non-entry points later or a way to specify that the files provided in the `input` option should not be considered entry points (and then we can mark the actual entry points later).

Let me know if I've misunderstood anything. I'd be happy to contribute to Rollup if we can iron out these details!

**Comment by lukastaegert** [comment]
> Rollup requires that input is specified in the input options

This is not true. The error message sounds like it but at the moment, you can already not supply `input` but emit a chunk in `buildStart` and Rollup will run without error. With the new hook, we could extend it so that it is also enough to just emit a chunk some time before `buildEnd`.

So the plugin could be like

```js
import globby from 'globby';
import { findEntryPoints } from 'find-entry-points'

function autoEntry(files) {
  let addedEntryPoints = false;

  return {
    buildStart() {
      for (const id of globby(files)) {
        this.addModule(id);
      }
    }

    modulesLoaded() {
      if (!addedEntryPoints) {
        addedEntryPoints = true;
        const entryPoints = (
        await findEntryPoints(this.getModuleIds(), {
          parseImports: ({ file: { path: moduleId } }) => {
            const { importedIds, dynamicallyImportedIds } = this.getModuleInfo(
              moduleId
            )
            return importedIds.concat(dynamicallyImportedIds)
          }
        })
      ).map(
        entryPointGroups =>
          entryPointGroups.find(
            entryPoint => basename(entryPoint) === `index.js`
          ) ?? entryPointGroups[0]
      );
        for (const id of entryPoints) {
          this.emitFile({type: 'chunk', id});
        }
      }
    }
  }
}
```

**Comment by frank-dspeed** [comment]
@lukastaegert +1 for your none entry file emit method that is exactly what is often needed i workaround that at present via supplying a single entrypoint a virtual one that imports the other entrypoints so they are not handled like entrypoints and then i remove the virtual one and then do manual chunk.

So long story short yes your idea is exactly the one that is needed then i do not need to expose this.graph :+1:

**Comment by lukastaegert** [comment]
If you close the issue it is harder to track it...

**Comment by TomerAberbach** [comment]
@lukastaegert, thanks for the clarification! For the moment I still think I'll use the `options` hook because it allows me to pass the cache from the nested Rollup instance to the top-level Rollup instance, but that's good to know.

Would you accept a PR to update the documentation website with this information? I think the website currently states that the `input` option in required.

**Comment by lukastaegert** [comment]
Certainly. I guess the point to clarify is that if the option is not supplied, a plugin needs to take care of emitting chunks.

**Comment by TomerAberbach** [comment]
> Certainly. I guess the point to clarify is that if the option is not supplied, a plugin needs to take care of emitting chunks.

Yup, will make sure to clarify that.

**Comment by dagda1** [comment]
I am wondering if this is something I can use.

I am writing a plugin to transform cucumber `*.feature` files in javascript esm modules.

the problem I have is that every `*.feature` file e.g.

```gherkin
Feature: Visit career guide page in career.guru99.com
Scenario: Visit career.guru99.com
Given: I browse to career.guru99.com
```

Has a backing js or ts file called a step definition

```javascript
import { Given, When, Then from 'cucumber';
import assert from 'assert';
import{ driver } from '../support/web_driver';

Given(/^browse to web site "([^"]*)"$/, async function(url) {
    return driver.get(url);
});

```

The problem I have is that there is no import statement for the backing `*.{js,ts}` file so rollup will not know about it.

Could I use a similar approach as is listed here to bring in the step definitions with no import statement anywhere?

**Comment by lukastaegert** [comment]
You can bring it in, but if it is not imported, it will not become part of the bundle unless you explicitly make it an entry point. Using `this.emitFile({type: 'chunk', id: stepDefinitionFileName})` could be what you are looking for.

**Comment by dagda1** [comment]
@lukeapage thank you still getting up to speed with rollup.  Where would this sit in the pipeline?  Would I have a plugin to find the files and write `this.emitFile` statements? 

Apologies for hijacking the thread.  

I did create an [SO question](https://stackoverflow.com/questions/63109484/how-to-use-rollup-to-parse-cucumber-feature-files-and-backing-step-definition-fi) if that is better
